# Population Genetics

I have read, that, given a population stuck on an island that reproduces between themselves for many generations and given enough generations, every individual from the original population will eventually either be an ancestor to **ALL** of the individuals or they will be an ancestor to **NONE** of the individuals.

I beleive the book I read it from was *The Selfish Gene* from Richard Dawkins where he argues that to find our common ancestors we don't necessarily have to go back all the way to the **one** individual that that is a common ancestor to all of us. We can go back a bit less to a handfull of ancestors who are ancestors to all of us.

This short code is an exploration to this idea. We simulate a population with N indiviuals each with X "genes". The original populations start out with there genes all equal to the individuals 'name' (number in our case). So as the generations go on, we count the types of genes to see how many of the original ancestors still contribute to the current population. Of these ancestors we count how many of the current population are decendent from him. And hopfully we will observe that eventually we will be left with only a handfull of the original genes that are present in 100% of the population.

In [1]:
import matplotlib.pyplot as plt
import collections
import numpy as np
import math
%matplotlib inline

## Reproduction

For reproduction we get at random half of the genes from one parent and half of the genes from the second parent and make a child out of them.

**Obs:** if there is an odd number of genes one of the parents contribuits more. But this doesn't really matter

In [2]:
def reproduce(parent_1, parent_2):
    assert len(parent_1) == len(parent_2)
    n = len(parent_1)
    half_floor = math.floor(n/2)
    half_ceil = math.ceil(n/2)
    genes_1 = np.random.choice(n,half_floor)
    genes_2 = np.random.choice(n,half_ceil)
    p1 = parent_1[genes_1]
    p2 = parent_2[genes_2]
    child = np.concatenate([p1, p2])
    return child

In [3]:
# example
parent_A = np.array([1,1,1,1,1,1])
parent_B = np.array([2,2,2,2,2,2])

child = reproduce(parent_A,parent_B)
print(child)

[1 1 1 2 2 2]


## Create the initial population

Here we choose the number of individuals in our initial population and the number of genes that each will have. Remember the initial individuals will start out with all their genes the same, this is what will help us track their decendents later on. So the initial gene of each individual is kind of Like their name, or their ID.

In [4]:
number_of_individuals = 100
number_of_genes = 100

names = list(range(number_of_individuals))

#creates the initial population
population = []
for name in names:
    new_guy = np.array([name for i in range(number_of_genes)])
    population.append(new_guy)

population = np.array(population)

## Generations

For each generation there are a couple of things to be done.
    1. Pair up the indiviuals
    2. Reproduce them
    3. Create new population
    4. Gather stats
Each of these could potentially have very intracate rules, but here we stick with the vary basics. So:

**1. Pair up the individuals:** we pair the population with a random permutaion of itself. This means that every individual reproduces twice (once as parent_1 and once as parent_2) per generation. This also implies that there are no sexes, any one can be paired up to anyone (including themselves). 

**2. Reproduce them:** the reproduciton was describe above. Basically a random half of one parents genes with a random half of the other parent's genes. Each coulpe produces only 1 child. Since each parent reproduces twice there are the same number of children as there are parents.

**3. Create new population:** We could potentially get a combination of parents and children, or make the population size vary, or do some tipe of selection. But, we do none of these. The new population is simply the children generated in the reproduction pahse.

**4. Gather stats:** There are a lot of interesting stats but we are only gathering 2. The number of original ancestors that still contribuite to the pool of genes, and The number of the current individuals which are decendents of each of the remaining ancestors.

In [5]:
number_of_generations = 2000
for epoch in range(number_of_generations):
    pop_size = len(population) 
    pair = np.random.permutation(pop_size)
    new_pop = []
    for i, individual in enumerate(population):
        child = reproduce(individual,population[pair[i]])
        new_pop.append(child)
    population = np.array(new_pop)
    stats = collections.Counter(population.flatten())
    num_ancestors = len(stats.keys())
    print('epoch',epoch, 'number of ancestors', num_ancestors)

epoch 0 number of ancestors 100
epoch 1 number of ancestors 100
epoch 2 number of ancestors 100
epoch 3 number of ancestors 100
epoch 4 number of ancestors 100
epoch 5 number of ancestors 100
epoch 6 number of ancestors 100
epoch 7 number of ancestors 100
epoch 8 number of ancestors 100
epoch 9 number of ancestors 100
epoch 10 number of ancestors 100
epoch 11 number of ancestors 100
epoch 12 number of ancestors 100
epoch 13 number of ancestors 100
epoch 14 number of ancestors 100
epoch 15 number of ancestors 100
epoch 16 number of ancestors 100
epoch 17 number of ancestors 100
epoch 18 number of ancestors 100
epoch 19 number of ancestors 100
epoch 20 number of ancestors 100
epoch 21 number of ancestors 100
epoch 22 number of ancestors 100
epoch 23 number of ancestors 100
epoch 24 number of ancestors 100
epoch 25 number of ancestors 100
epoch 26 number of ancestors 100
epoch 27 number of ancestors 100
epoch 28 number of ancestors 100
epoch 29 number of ancestors 100
epoch 30 number of a

epoch 256 number of ancestors 54
epoch 257 number of ancestors 54
epoch 258 number of ancestors 54
epoch 259 number of ancestors 54
epoch 260 number of ancestors 54
epoch 261 number of ancestors 54
epoch 262 number of ancestors 54
epoch 263 number of ancestors 54
epoch 264 number of ancestors 54
epoch 265 number of ancestors 54
epoch 266 number of ancestors 54
epoch 267 number of ancestors 54
epoch 268 number of ancestors 54
epoch 269 number of ancestors 54
epoch 270 number of ancestors 54
epoch 271 number of ancestors 54
epoch 272 number of ancestors 54
epoch 273 number of ancestors 54
epoch 274 number of ancestors 54
epoch 275 number of ancestors 54
epoch 276 number of ancestors 54
epoch 277 number of ancestors 54
epoch 278 number of ancestors 54
epoch 279 number of ancestors 54
epoch 280 number of ancestors 54
epoch 281 number of ancestors 54
epoch 282 number of ancestors 54
epoch 283 number of ancestors 54
epoch 284 number of ancestors 54
epoch 285 number of ancestors 54
epoch 286 

epoch 510 number of ancestors 31
epoch 511 number of ancestors 31
epoch 512 number of ancestors 30
epoch 513 number of ancestors 30
epoch 514 number of ancestors 30
epoch 515 number of ancestors 30
epoch 516 number of ancestors 30
epoch 517 number of ancestors 30
epoch 518 number of ancestors 30
epoch 519 number of ancestors 30
epoch 520 number of ancestors 30
epoch 521 number of ancestors 30
epoch 522 number of ancestors 29
epoch 523 number of ancestors 29
epoch 524 number of ancestors 29
epoch 525 number of ancestors 29
epoch 526 number of ancestors 29
epoch 527 number of ancestors 29
epoch 528 number of ancestors 29
epoch 529 number of ancestors 29
epoch 530 number of ancestors 29
epoch 531 number of ancestors 29
epoch 532 number of ancestors 29
epoch 533 number of ancestors 29
epoch 534 number of ancestors 29
epoch 535 number of ancestors 29
epoch 536 number of ancestors 29
epoch 537 number of ancestors 29
epoch 538 number of ancestors 29
epoch 539 number of ancestors 29
epoch 540 

epoch 760 number of ancestors 19
epoch 761 number of ancestors 19
epoch 762 number of ancestors 19
epoch 763 number of ancestors 19
epoch 764 number of ancestors 19
epoch 765 number of ancestors 19
epoch 766 number of ancestors 19
epoch 767 number of ancestors 19
epoch 768 number of ancestors 19
epoch 769 number of ancestors 19
epoch 770 number of ancestors 19
epoch 771 number of ancestors 19
epoch 772 number of ancestors 19
epoch 773 number of ancestors 19
epoch 774 number of ancestors 19
epoch 775 number of ancestors 19
epoch 776 number of ancestors 19
epoch 777 number of ancestors 19
epoch 778 number of ancestors 19
epoch 779 number of ancestors 19
epoch 780 number of ancestors 19
epoch 781 number of ancestors 19
epoch 782 number of ancestors 19
epoch 783 number of ancestors 19
epoch 784 number of ancestors 19
epoch 785 number of ancestors 19
epoch 786 number of ancestors 19
epoch 787 number of ancestors 19
epoch 788 number of ancestors 19
epoch 789 number of ancestors 19
epoch 790 

epoch 1011 number of ancestors 14
epoch 1012 number of ancestors 14
epoch 1013 number of ancestors 14
epoch 1014 number of ancestors 14
epoch 1015 number of ancestors 14
epoch 1016 number of ancestors 14
epoch 1017 number of ancestors 14
epoch 1018 number of ancestors 14
epoch 1019 number of ancestors 14
epoch 1020 number of ancestors 14
epoch 1021 number of ancestors 14
epoch 1022 number of ancestors 14
epoch 1023 number of ancestors 14
epoch 1024 number of ancestors 14
epoch 1025 number of ancestors 14
epoch 1026 number of ancestors 14
epoch 1027 number of ancestors 14
epoch 1028 number of ancestors 14
epoch 1029 number of ancestors 14
epoch 1030 number of ancestors 14
epoch 1031 number of ancestors 14
epoch 1032 number of ancestors 14
epoch 1033 number of ancestors 14
epoch 1034 number of ancestors 14
epoch 1035 number of ancestors 14
epoch 1036 number of ancestors 14
epoch 1037 number of ancestors 14
epoch 1038 number of ancestors 14
epoch 1039 number of ancestors 14
epoch 1040 num

epoch 1258 number of ancestors 14
epoch 1259 number of ancestors 14
epoch 1260 number of ancestors 14
epoch 1261 number of ancestors 14
epoch 1262 number of ancestors 14
epoch 1263 number of ancestors 14
epoch 1264 number of ancestors 14
epoch 1265 number of ancestors 14
epoch 1266 number of ancestors 14
epoch 1267 number of ancestors 14
epoch 1268 number of ancestors 14
epoch 1269 number of ancestors 14
epoch 1270 number of ancestors 14
epoch 1271 number of ancestors 14
epoch 1272 number of ancestors 14
epoch 1273 number of ancestors 14
epoch 1274 number of ancestors 14
epoch 1275 number of ancestors 14
epoch 1276 number of ancestors 14
epoch 1277 number of ancestors 14
epoch 1278 number of ancestors 14
epoch 1279 number of ancestors 14
epoch 1280 number of ancestors 14
epoch 1281 number of ancestors 14
epoch 1282 number of ancestors 14
epoch 1283 number of ancestors 14
epoch 1284 number of ancestors 14
epoch 1285 number of ancestors 14
epoch 1286 number of ancestors 14
epoch 1287 num

epoch 1501 number of ancestors 13
epoch 1502 number of ancestors 13
epoch 1503 number of ancestors 13
epoch 1504 number of ancestors 13
epoch 1505 number of ancestors 13
epoch 1506 number of ancestors 13
epoch 1507 number of ancestors 13
epoch 1508 number of ancestors 13
epoch 1509 number of ancestors 13
epoch 1510 number of ancestors 13
epoch 1511 number of ancestors 13
epoch 1512 number of ancestors 13
epoch 1513 number of ancestors 13
epoch 1514 number of ancestors 13
epoch 1515 number of ancestors 13
epoch 1516 number of ancestors 13
epoch 1517 number of ancestors 13
epoch 1518 number of ancestors 13
epoch 1519 number of ancestors 13
epoch 1520 number of ancestors 13
epoch 1521 number of ancestors 13
epoch 1522 number of ancestors 13
epoch 1523 number of ancestors 13
epoch 1524 number of ancestors 13
epoch 1525 number of ancestors 13
epoch 1526 number of ancestors 13
epoch 1527 number of ancestors 13
epoch 1528 number of ancestors 13
epoch 1529 number of ancestors 13
epoch 1530 num

epoch 1744 number of ancestors 11
epoch 1745 number of ancestors 11
epoch 1746 number of ancestors 11
epoch 1747 number of ancestors 11
epoch 1748 number of ancestors 11
epoch 1749 number of ancestors 11
epoch 1750 number of ancestors 11
epoch 1751 number of ancestors 11
epoch 1752 number of ancestors 11
epoch 1753 number of ancestors 11
epoch 1754 number of ancestors 11
epoch 1755 number of ancestors 11
epoch 1756 number of ancestors 11
epoch 1757 number of ancestors 11
epoch 1758 number of ancestors 11
epoch 1759 number of ancestors 11
epoch 1760 number of ancestors 11
epoch 1761 number of ancestors 11
epoch 1762 number of ancestors 11
epoch 1763 number of ancestors 11
epoch 1764 number of ancestors 11
epoch 1765 number of ancestors 11
epoch 1766 number of ancestors 11
epoch 1767 number of ancestors 11
epoch 1768 number of ancestors 11
epoch 1769 number of ancestors 11
epoch 1770 number of ancestors 11
epoch 1771 number of ancestors 11
epoch 1772 number of ancestors 11
epoch 1773 num

epoch 1986 number of ancestors 11
epoch 1987 number of ancestors 11
epoch 1988 number of ancestors 11
epoch 1989 number of ancestors 11
epoch 1990 number of ancestors 11
epoch 1991 number of ancestors 11
epoch 1992 number of ancestors 11
epoch 1993 number of ancestors 11
epoch 1994 number of ancestors 11
epoch 1995 number of ancestors 11
epoch 1996 number of ancestors 11
epoch 1997 number of ancestors 11
epoch 1998 number of ancestors 11
epoch 1999 number of ancestors 11


## Number of Ancestors

In [10]:
print('Final number of Ancestors:',len(stats))

Final number of Ancestors: 11


So from the 100 original individuals only 11 continue to contribute to the pool of genes after 200 generations!

## Decendents

Lets look at how many of the individuals each of the ancestors are present in

In [7]:
descendents_count = {ind:0 for ind in stats.keys()}
for ind in population:
    for gene in stats.keys():
        if gene in ind: 
            descendents_count[gene] +=1
descendents_count

{15: 99,
 27: 39,
 39: 98,
 54: 100,
 56: 100,
 58: 100,
 60: 85,
 61: 100,
 67: 98,
 89: 100,
 95: 100}

Out of the 11 remaining ancestors, **6** are common acestors to ALL of the new individuals. **9** are ancestors to at least **98%** of the population. 

## Conclusions

The effect we observed is not quite as dramatic as the claims of the book. But they are pretty impressive none the less. Its also important to note that **100 individuals** and **100 genes** each is a vary small size compared to the actual number of people and genes in the world. And **2000 generations** considering that each generation is seperated from the last by 20 years would mean that we did 40000 years of inbreeding of 100 individuals. Not exactly a propable situation.

I invite you to play around with the number of individuals, the number of genes and the number of generations. If you are feeling especially adventurous play around also with the size of the population over time.